hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -32,25 +35,43 @@ func NewStorageTypeFactory(group string, idFactory astmodel.IdentifierFactory) *
 
 	result := &StorageTypeFactory{
 		group:             group,
-		inputTypes:        types,
+		referenceTypes:    types,
+		outputTypes:       make(astmodel.Types),
+		specTypes:         astmodel.NewTypeNameSet(),
+		statusTypes:       astmodel.NewTypeNameSet(),
 		idFactory:         idFactory,
 		conversionMap:     make(map[astmodel.PackageReference]astmodel.PackageReference),
 		functionInjector:  NewFunctionInjector(),
 		resourceHubMarker: NewHubVersionMarker(),
-		typeConverter:     NewTypeConverter(types),
+		typeConverter:     NewTypeConverter(types, idFactory),
 	}
 
 	return result
 }
 
 // Add the supplied type definition to this factory
-func (f *StorageTypeFactory) Add(def astmodel.TypeDefinition) {
-	f.inputTypes.Add(def)
+func (f *StorageTypeFactory) Add(definitions ...astmodel.TypeDefinition) {
+	for _, def := range definitions {
+		f.referenceTypes.Add(def)
+
+		if rt, ok := astmodel.AsResourceType(def.Type()); ok {
+			// We have a resource type
+			if tn, ok := astmodel.AsTypeName(rt.SpecType()); ok {
+				// Keep track of all our spec types
+				f.specTypes = f.specTypes.Add(tn)
+			}
+			if tn, ok := astmodel.AsTypeName(rt.StatusType()); ok {
```

Comment:
minor: Would a resource without a spec sneaking in here cause problems? Would it be worth failing-fast in that case (could `panic` if you don't want to pollute this method signature?).

Ignore this if there's no issue and a spec-less resource isn't going to blow up later in this stage.

Reply:
Can we have a resource without a spec? What would that even mean? There'd be nothing to submit to ARM, right?

Reply:
> Can we have a resource without a spec?

Without a bug, no I don't think we can. My question here was basically, you're doing these two checks for spec and status and if somehow there was no spec and no status this loop wouldn't error it'd just skip adding anything - in other places I've seen this code there's a `!ok` case on the spec path that blows up if there's no spec. Did you want that here, or is it really not important in this context and if we wanted to assert on that we should do it elsewhere elsewhere?

I'm fine with either answer.

Reply:
If we have a resource with no spec, I think it will blow up before we get here. I don't see the need to check.